transport: pass network channel exceptions to close listeners

schase-es opened the pull request:
Previously, exceptions encountered on a netty channel were caught and logged at some level, but not passed to the TcpChannel or Transport.Connection close listeners. This limited observability. This change implements this exception reporting and passing, with TcpChannel.onException and NodeChannels.closeAndFail reporting exceptions and their close listeners receiving them. Some test infrastructure (FakeTcpChannel) and assertions in close listener onFailure methods have been updated.

Closes: ES-11644

schase-es committed:
transport: pass network channel exceptions to close listeners

Previously, exceptions encountered on a netty channel were caught and logged at
some level, but not passed to the TcpChannel or Transport.Connection close
listeners. This limited observability. This change implements this exception
reporting and passing, with TcpChannel.onException and NodeChannels.closeAndFail
reporting exceptions and their close listeners receiving them. Some test
infrastructure (FakeTcpChannel) and assertions in close listener onFailure
methods have been updated.

Closes: ES-11644

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)

schase-es commented:
As discussed (https://github.com/elastic/elasticsearch/pull/127736), this is the first section broken into its own PR. I put some work into adding testing. Otherwise, there was one deleted line that was unnecessary, and a small comment I added to the Netty4TcpChannel.

DaveCTurner commented on modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java:
```diff
@@ -383,12 +396,12 @@ protected InboundPipeline getInboundPipeline(Channel ch, boolean isRemoteCluster
-            if (channelFuture.isSuccess() == false && logger.isDebugEnabled()) {
-                logger.debug(format("exception while closing channel: %s", channelFuture.channel()), channelFuture.cause());
+    private static void addClosedExceptionLogger(Netty4TcpChannel channel) {
+        channel.addCloseListener(ActionListener.wrap((ignored) -> {}, (e) -> {
```
AIUI this is a change of behaviour right? `channel.closeFuture().isSuccess()` is never false so in practice this doesn't do anything today, whereas now we're going to log the exception that caused the channel to close. But I think we already log such exceptions, so maybe we don't need to duplicate that logging here.

Maybe we should just delete this dead code?

Also `ActionListener#wrap` isn't really necessary since the `onResponse` can't throw anything. If we do want to keep this behaviour it'd be better not to instantiate those two extra lambdas unnecessarily.

DaveCTurner commented on server/src/test/java/org/elasticsearch/transport/InboundHandlerTests.java:
```diff
@@ -255,6 +259,11 @@ public void testClosesChannelOnErrorInHandshake() throws Exception {
             handler.inboundMessage(channel, requestMessage);
             assertTrue(isClosed.get());
+            assertTrue(closeListener.isDone());
+            try {
+                closeListener.get();
+                assert false : "channel should have an exception reported";
+            } catch (Exception e) {}
```
Suggest `expectThrows` for this pattern.

DaveCTurner commented on server/src/main/java/org/elasticsearch/transport/TcpChannel.java:
```diff
@@ -66,6 +66,13 @@ public interface TcpChannel extends CloseableChannel {
+     *
+     * @param e the exception
+     */
+    void onException(Exception e);
```
Naming nit, maybe `setCloseException` or something like that? To me `onException` sounds like something that should react to the exception by e.g. closing the channel and that's not what is happening.

DaveCTurner commented on modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java:
```diff
@@ -337,17 +344,23 @@ protected ServerChannelInitializer(String name) {
 
         @Override
         public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+            Netty4TcpChannel channel = ctx.channel().attr(CHANNEL_KEY).get();
+            if (cause instanceof Error) {
+                channel.onException(new Exception(cause));
+            } else {
+                channel.onException((Exception) cause);
+            }
```
Repeated code, probably best to extract a method?

DaveCTurner commented on modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4TcpChannel.java:
```diff
@@ -34,15 +34,25 @@ public class Netty4TcpChannel implements TcpChannel {
     private final ListenableFuture<Void> closeContext = new ListenableFuture<>();
     private final ChannelStats stats = new ChannelStats();
     private final boolean rstOnClose;
+    private volatile Exception channelError = null;
```
Naming nit: `channelException`? `Error` implies something fatal to the node.

DaveCTurner commented on modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4TcpChannel.java:
```diff
@@ -34,15 +34,25 @@ public class Netty4TcpChannel implements TcpChannel {
         this.rstOnClose = rstOnClose;
-        addListener(this.channel.closeFuture(), closeContext);
         addListener(connectFuture, connectContext);
+        // The netty closeFuture is programmed to never return an exception.
+        // This listener takes close-causing exceptions reported during the
+        // channel lifetime, and reports them to the closeListener.
```
Nit: I'd be inclined not to add comments like this, it's just repeating what the code says (and risks diverging from the code over time). But consider adding a javadoc to the `channelError` field. Potentially also an assertion that `closeFuture` is never completed exceptionally rather than simply stating it in a comment, that way the reader can see that it's really true.

DaveCTurner commented on server/src/main/java/org/elasticsearch/transport/TcpTransport.java:
```diff
@@ -1181,7 +1205,16 @@ public void onResponse(Void ignored) {
         public void onFailure(Exception e) {
-            assert false : e; // never called
+            long closeTimeMillis = threadPool.relativeTimeInMillis();
+            logger.info(
```
I think I'd rather keep this at `debug` for now.

DaveCTurner commented on server/src/test/java/org/elasticsearch/transport/TcpTransportTests.java:
```diff
@@ -606,7 +605,10 @@ private void testExceptionHandling(
-                assertThat(listener.actionGet(), nullValue());
+                try {
+                    listener.get();
+                    assert false : "channel should have an exception reported";
```
Can we have a test that sees the channel closed without an exception too?

DaveCTurner reviewed (commented):
All makes sense to me, I left some comments but nothing major.

schase-es committed:
Addressed review feedback:
- renamed channelError and onException
- removed dead channel error logging code
- used expectThrows pattern in tests
- added closeListener test
- added assert to onFailure branch of netty channel future/listener adapter
- de-duplicated throwable/exception adapting code
- log transport errors at debug

schase-es commented:
I addressed all the straightforward things -- thanks for the feedback :)

I had a hard time getting a test into TcpTranportTests that wasn't a trivial test of a FakeTcpChannel. I was able to find a place in ClusterConnectionManagerTests that looked fine, but is at the Transport.Connection level.

Please comment with any additional ideas -- the test infrastructure in this area is a lot to take in.

DaveCTurner commented on server/src/main/java/org/elasticsearch/transport/TcpTransport.java:
```diff
@@ -1181,7 +1205,16 @@ public void onResponse(Void ignored) {
+            long closeTimeMillis = threadPool.relativeTimeInMillis();
+            logger.debug(
+                () -> format(
+                    "closed transport connection [{}] to [{}] with age [{}ms], exception:",
```
You need `%s` rather than `{}` when using `Strings.format` to format a string:

```
[2025-05-09T08:58:55,854][DEBUG][o.e.t.TcpTransport       ][node_t3][transport_worker][T#1] closed transport connection [{}] to [{}] with age [{}ms], exception:
java.net.SocketException: Connection reset
    at sun.nio.ch.SocketChannelImpl.throwConnectionReset(SocketChannelImpl.java:399) ~[?:?]
    at sun.nio.ch.SocketChannelImpl.implRead(SocketChannelImpl.java:431) ~[?:?]
    at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:489) ~[?:?]
    at org.elasticsearch.transport.netty4.CopyBytesSocketChannel.readFromSocketChannel(CopyBytesSocketChannel.java:132) ~[main/:?]
    at org.elasticsearch.transport.netty4.CopyBytesSocketChannel.doReadBytes(CopyBytesSocketChannel.java:117) ~[main/:?]
    at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:151) ~[netty-transport-4.1.118.Final.jar:4.1.118.Final]
    at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796) ~[netty-transport-4.1.118.Final.jar:4.1.118.Final]
    at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:697) ~[netty-transport-4.1.118.Final.jar:4.1.118.Final]
    at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:660) ~[netty-transport-4.1.118.Final.jar:4.1.118.Final]
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[netty-transport-4.1.118.Final.jar:4.1.118.Final]
    at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998) ~[netty-common-4.1.118.Final.jar:4.1.118.Final]
    at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.118.Final.jar:4.1.118.Final]
    at java.lang.Thread.run(Thread.java:1447) ~[?:?]
```

This catches me out from time to time too 😁

DaveCTurner reviewed (commented):
Yeah good point on the testing, it's a little tricky. The only observable change here AFAICT is how the `ChannelCloseLogger` logs the exception now if there is an exception, so I think that's what I'd try. We test this logging in `org.elasticsearch.transport.netty4.ESLoggingHandlerIT#testConnectionLogging` today. Can we use that? It seems empirically that simply starting and stopping a node yields some clean closes and some `Connection Reset` exceptions.

This isn't really the right place for that test, we should create a `TcpTransportIT` and move it there too.

schase-es commented on modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java:
```diff
@@ -308,18 +307,27 @@ protected void stopInternal() {
         }, serverBootstraps::clear, () -> clientBootstrap = null);
     }
 
+    private Exception exceptionFromThrowable(Throwable cause) {
```
Part of the reasoning for separating just this bit out, was that this pattern is potentially widely used enough to be in ExceptionsHelper. There's actually another exception handler at the bottom of this file that uses this, and it's in several other netty error receivers.

I also thought about making the TcpChannel setCloseException take a throwable instead, and do this internally.

Most of these exception handlers are short and have the same 2-3 things, depending on the context. I think this is okay for now?

schase-es committed:
Addressed next round of review feedback:
- added test to check for close logging, with and without exception
- fixed up close logger formatting
- javadoc fixup for channel exception field
- further refactored remaining code with throwable -> exception in Netty4Transport

DaveCTurner commented on modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/transport/netty4/ESLoggingHandlerIT.java:
```diff
@@ -105,4 +112,46 @@ public void testConnectionLogging() throws IOException {
+        int failAttempts = 0;
+        do {
+            internalCluster().restartNode(nodeName);
+        } while (latch.await(500, TimeUnit.MILLISECONDS) == false && failAttempts++ < 10);
```
Did you encounter cases where we don't get an exceptional close on the first try?

schase-es replied:
Hi David -- great question. This turned out to not be needed, so I've removed it. I think this grew out of an earlier version that was broken, and got lost in some debugging and breakpoints.

nicktindall reviewed (commented):
LGTM, pending @DaveCTurner's outstanding comments. Great work.

schase-es committed:
Addressed last review feedback:
- removed complicated do... while logic around network exceptions on close

schase-es committed:
Merge branch 'main' into ES-11644_channel-exception-passing

DaveCTurner reviewed (approved):
LGTM

For possible future reference: I am not 100% convinced that restarting a node will always yield both clean and exceptional closes of transport connections as asserted in `ESLoggingHandlerIT` with this change. Empirically it seems to be true, but CI will tell us for sure whether we need to do something more elaborate.
